### DISCUSSION TITLE
How to simulate browser to keepalive socket.io-client use code ？

### DISCUSSION THREAD

**Discussion by lajiao1874** [socketio/socket.io#4630] [Q&A] [answered]
How to simulate browser to keepalive socket.io-client use code ？

**Comment by darrachequesne** [comment, marked as answer]
Hi! I'm not sure to understand your use case. Could you please explain what you mean by "keepalive"?

**Reply by lajiao1874** [reply]
Sorry.
I mean that i want to simulate browser send ping message to socket.io-server in my script,。
Do we have a ready way to do it ?

thanks

**Reply by qiulang** [reply]
But the ping/pong keeyalive message is sent by socket.io-client automatically, so what do you mean "simulate browser send ping message" ?

**Reply by darrachequesne** [reply]
@liyoung1874 you should be able to open a plain WebSocket connection, and send a "ping" event to the your server:

```js
const socket = new WebSocket("ws://localhost:3000/socket.io/?EIO=4&transport=websocket");

socket.addEventListener("message", () => {
  // the Engine.IO handshake
  console.log(data); 
  // send a Socket.IO CONNECT packet
  socket.send("40");
  // send a "ping" event
  socket.send('42["ping"]');
}, { once: true });
```

Note: if you want to keep the connection open, you will have to respond to "ping" packets sent by the server (aka. the heartbeat mechanism)

References:

- https://socket.io/fr/docs/v4/engine-io-protocol/
- https://socket.io/fr/docs/v4/socket-io-protocol/
- https://github.com/socketio/socket.io-protocol/blob/main/test-suite/test-suite.js

**Reply by lajiao1874** [reply]
Thanks very much !

**Reply by qiulang** [reply]
@darrachequesne https://socket.io/docs/v4/ has these words "Although Socket.IO indeed uses WebSocket for transport when possible, it adds additional metadata to each packet. That is why a **WebSocket client will not be able to successfully connect to a Socket.IO server**, and a Socket.IO client will not be able to connect to a plain WebSocket server either."

So I am a bit confused with your answer. Does that mean a plain WebSocket connection can only ping/pong but can not actually send message data or does it mean with ping/pong a a plain WebSocket connection can actually connect to Socket.IO server and send message to it (with **4** at the beginning of message) ?

**Reply by darrachequesne** [reply]
@qiulang that means that you cannot just use a WebSocket client like that:

```js
const socket = new WebSocket("ws://example.com");

socket.addEventListener("open", () => {
  socket.emit("hello");
});
```

There are a few constraints that the WebSocket client has to respect:

- include the query parameters

```js
const socket = new WebSocket("ws://example.com/socket.io/?EIO=4&transport=websocket");
```

- respond to ping packets from the server

```js
socket.addEventListener("message", ({ data }) => {
  if (data === "2") {
    socket.send("3");
  }
});
```

- send a CONNECT packet to reach a namespace

```js
// main namespace
socket.send("40");
// custom namespace
socket.send("40/admin,");
```

- properly serialize the payload

```js
// socket.emit("hello", "world") with the Socket.IO client becomes
socket.send('42["hello","world"]');
```

If the client does not respect one of those rules, then the connection will be closed by the server.

Hope that clarify things!

**Reply by qiulang** [reply]
Yes very much (and just like what I thought)!
